linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -673,19 +678,43 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
   private def transformClosureCommon(arrow: Boolean,
       captureParams: List[ParamDef], params: List[ParamDef],
       restParam: Option[ParamDef], body: Tree,
-      newCaptureValues: List[Tree])(
-      implicit scope: Scope, pos: Position): Closure = {
+      tcaptureValues: List[PreTransform])(cont: PreTransCont)(
+      implicit scope: Scope, pos: Position): TailRec[Tree] = {
 
     val thisType = if (arrow) NoType else AnyType
-    val (allNewParams, newBody) = transformIsolatedBody(None, thisType,
-        captureParams ++ params ++ restParam, AnyType, body, scope.implsBeingInlined)
-    val (newCaptureParams, newParams, newRestParam) = {
-      val (c, t) = allNewParams.splitAt(captureParams.size)
-      if (restParam.isDefined) (c, t.init, Some(t.last))
-      else (c, t, None)
+
+    val (allNewParamsWithUse, newBody) =
+      transformIsolatedBody(None, thisType, captureParams ++ params ++ restParam, AnyType, body,
+        scope.implsBeingInlined)
+
+    val (newCaptureParamsWithUse, newParamsWithUse) = allNewParamsWithUse.splitAt(captureParams.size)
+
+    val (newParams, newRestParam) = {
+      val t = newParamsWithUse.map(_._1).toList
+      if (restParam.isDefined) (t.init, Some(t.last))
+      else (t, None)
     }
 
-    Closure(arrow, newCaptureParams, newParams, newRestParam, newBody, newCaptureValues)
+    val captureBindings =
+      newCaptureParamsWithUse.iterator.zip(tcaptureValues.iterator).map { case ((param, _), value) =>
+        Binding.temp(param.name.name, param.ptpe, mutable = false, value)
+      }.toList
+
+    withNewLocalDefs(captureBindings)((localDefs, cont1) => {
+      val newClosure = {
+        val finalCaptureParams = List.newBuilder[ParamDef]
+        val finalCaptureValues = List.newBuilder[Tree]
+        localDefs.iterator.zip(newCaptureParamsWithUse.iterator).foreach { case (localDef, (param, use)) =>
+          if (use.isUsed) {
+            finalCaptureParams += param
+            finalCaptureValues += localDef.newReplacement
+          }
+        }
```

Comment:
Use a for loop?

```scala
val (finalCaptureParams, finalCaptureValues) = (for {
  (localDef, (param, use) <- localDefs.iterator.zip(newCaptureParamsWithUse.iterator)
  if use.isUsed
} yield {
  param -> localDef.newReplacement
}).unzip
```

Reply:
Ok np will do but I need you to know you're causing my physical pain my making me perform unnecessary allocations 😅😭